src/main/java/org/gradle/crypto/checksum/Checksum.java

```diff
@@ -19,22 +19,30 @@
 import com.google.common.hash.HashFunction;
 import com.google.common.hash.Hashing;
 import com.google.common.io.Files;
-import org.gradle.api.Action;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.GradleException;
-import org.gradle.api.file.ConfigurableFileTree;
 import org.gradle.api.file.FileCollection;
-import org.gradle.api.tasks.Input;
-import org.gradle.api.tasks.InputFiles;
-import org.gradle.api.tasks.OutputDirectory;
-import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.file.FileSystemOperations;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.tasks.*;
 import org.gradle.api.tasks.incremental.IncrementalTaskInputs;
-import org.gradle.api.tasks.incremental.InputFileDetails;
+import org.gradle.util.GradleVersion;
 
+import javax.inject.Inject;
 import java.io.File;
 import java.io.IOException;
 
+@CacheableTask
 public class Checksum extends DefaultTask {
+
+    interface FileOps {
+        @Inject
+        FileSystemOperations getFs();
+    }
+
+    private final ObjectFactory objectFactory;
+    private FileOps fileOps;
+
     private FileCollection files;
     private File outputDir;
     private Algorithm algorithm;
```

Comment:
This could be turned into proper properties. This would require keeping getters/setters that delegate to the new properties for backwards compatibility.
```
DirectoryProperty outputDirectory
Property<Algorithm> checksumAlgorithm
```
Not required for this PR

Reply:
What's the benefit of using "true" properties?

Reply:
Lazy configuration https://docs.gradle.org/current/userguide/lazy_configuration.html

One example is what would happen if something changes `project.buildDir` after this task is registered.
`outputDirectory` should use `projectLayout.buildDirectory.dir("checksums")` as its convention.

Reply:
that's interesting, I'll apply it!